c2rust-analyze/src/rewrite/expr/unlower.rs

```diff
@@ -0,0 +1,379 @@
+use crate::rewrite::span_index::SpanIndex;
+use log::*;
+use rustc_hir as hir;
+use rustc_hir::intravisit;
+use rustc_hir::HirId;
+use rustc_middle::hir::nested_filter;
+use rustc_middle::mir::{self, Body, Location};
+use rustc_middle::ty::{DefIdTree, TyCtxt};
+use rustc_span::symbol::sym;
+use rustc_span::Span;
+use std::collections::btree_map::{BTreeMap, Entry};
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug, Hash)]
+pub struct MirOrigin {
+    pub hir_id: HirId,
+    pub span: Span,
+    pub desc: MirOriginDesc,
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug, Hash)]
+pub enum MirOriginDesc {
+    /// This MIR represents the whole HIR expression.
+    Expr,
+    /// This MIR stores the result of the HIR expression into a MIR local of some kind.
+    StoreIntoLocal,
+}
+
+struct UnlowerVisitor<'a, 'tcx> {
+    tcx: TyCtxt<'tcx>,
+    mir: &'a Body<'tcx>,
+    span_index: SpanIndex<Location>,
+    /// Maps MIR (sub)locations to the HIR node that produced each one, if known.
+    mir_map: BTreeMap<(Location, Vec<SubLoc>), MirOrigin>,
+}
+
+impl<'a, 'tcx> UnlowerVisitor<'a, 'tcx> {
+    fn location_span(&self, loc: Location) -> Span {
+        self.mir
+            .stmt_at(loc)
+            .either(|stmt| stmt.source_info.span, |term| term.source_info.span)
+    }
+
+    fn record(&mut self, loc: Location, sub_loc: &[SubLoc], ex: &hir::Expr) {
+        self.record_desc(loc, sub_loc, ex, MirOriginDesc::Expr);
+    }
+
+    fn record_desc(
+        &mut self,
+        loc: Location,
+        sub_loc: &[SubLoc],
+        ex: &hir::Expr,
+        desc: MirOriginDesc,
+    ) {
+        let origin = MirOrigin {
+            hir_id: ex.hir_id,
+            span: ex.span,
+            desc,
+        };
+        match self.mir_map.entry((loc, sub_loc.to_owned())) {
+            Entry::Vacant(e) => {
+                e.insert(origin);
+            }
+            Entry::Occupied(e) => {
+                if *e.get() != origin {
+                    let old_origin = e.get().clone();
+                    error!(
+                        "conflicting origins for {:?} {:?} ({:?})\n\
+                            origin 1 = {:?}\n\
+                            origin 2 = {:?}\n\
+                            expr 2 = {:?}",
+                        loc,
+                        sub_loc,
+                        self.location_span(loc),
+                        origin,
+                        old_origin,
+                        ex,
+                    );
+                }
+            }
+        }
+    }
+
+    fn get_sole_assign(
+        &self,
+        locs: &[Location],
+    ) -> Option<(Location, mir::Place<'tcx>, &'a mir::Rvalue<'tcx>)> {
+        if locs.len() != 1 {
+            return None;
+        }
+        self.get_last_assign(locs)
+    }
+
+    fn get_last_assign(
+        &self,
+        locs: &[Location],
+    ) -> Option<(Location, mir::Place<'tcx>, &'a mir::Rvalue<'tcx>)> {
+        let &loc = locs.last()?;
```

Comment:
```suggestion
        let loc = *locs.last()?;
```

The `&` destructure in a normal `let` is rarely used I think, so `*` would IMO be a lot clearer.